exercises/nondet_monad_cont/descr.md

```diff
@@ -0,0 +1,521 @@
+
+Because of this remark, one might be tempted to define the type `'a
+m` as a synonym for `'a Seq.t`, the type of sequences of values of
+type `'a`. However, although it is possible to represent a
+computation internally as a sequence (and we will do so in this
+exercise), this is not necessarily the best implementation
+technique. Thus, we prefer to make the API more flexible by viewing
+`'a m` as an abstract type and by offering an observation function,
+`sols`, which converts a computation to a sequence.
+
+## The Nondeterminism Monad's API
+
+The signature, or API, of the nondeterminism monad is as follows:
+
+```
+  (* Type. *)
+  type 'a m
+
+  (* Constructor functions. *)
+  val return: 'a -> 'a m
+  val (>>=): 'a m -> ('a -> 'b m) -> 'b m
+  val fail: 'a m
+  val choose: 'a m -> 'a m -> 'a m
+  val delay: (unit -> 'a m) -> 'a m
+  val at_most_once: 'a m -> 'a m
+  val interleave: 'a m -> 'a m -> 'a m
+  val (>>-): 'a m -> ('a -> 'b m) -> 'b m
+
+  (* Observation function. *)
+  val sols: 'a m -> 'a Seq.t
+```
+
+As explained above, a value of type `'a m` is **a description of a
+computation**, which, once executed, produces a sequence of results
+of type `'a`.
+
+To execute a computation `m`, one must first convert it to a
+sequence of type `'a Seq.t`, whose elements can then be demanded,
+one by one. (More information on the module `Seq` is given below.)
+This conversion is performed by the observation function `sols`.
+
+The call `sols m` typically terminates in constant time; the actual
+computation described by `m` takes place only when the elements of
+the sequence `sols m` are demanded, and only insofar as necessary to
+produce the elements that are demanded. For instance, applying
+`Seq.head` to the sequence `sols m` forces the computation to
+proceed up to the point where it is able to produce its first
+result.
+
+The constructor functions `return` and `(>>=)` exist in all monads.
+(They are also known as `return` and `bind`.) `return` constructs a
+trivial computation, which does nothing except return a value,
+whereas `(>>=)` constructs the sequential composition of two
+computations. Together, they allow constructing the sequential
+composition of an arbitrary number of computations.
+
+* The computation `return v` succeeds exactly once with the value
+  `v`. In other words, the sequence of values that it produces is
+  the singleton sequence composed of just `v`.
+
+* The computation `m1 >>= m2` is the sequential composition of the
+  computations `m1` and `m2`. This composition operator is
+  asymmetric: whereas its first argument `m1` is a computation of
+  type `'a m`, its second argument `m2` is a function of type `a ->
+  'b m`. Every value `x` produced by `m1` is passed to `m2`,
+  yielding a computation `m2 x`. The sequence of values produced by
+  `m1 >>= m2` is the concatenation of the sequences of values
+  produced by the computations `m2 x`, where `x` ranges over the
+  values produced by `m1`.
+
```

Comment:
The reader might be wondering if the ordering of values is important. I guess we should add that it is not specified.

Reply:
Yes, order matters. (For instance, `at_most_once` selects only the *first* result.) The text says "sequence" everywhere and never "set", so I think it should be clear that order matters.